### ISSUE TITLE
Issue with query Alias in Materialized View 

### ISSUE THREAD

**Issue by TH-HA** [ClickHouse/ClickHouse#3450]
Hello 
I'm currently using CH 18.14.8 and i'm facing an issue when building materialized views.
I'm trying to build and feed automatically a site aggregated table from an input table contain site source and destination.  As union is not supported in MV queries , i workaround the pbm using arrays. 
But i'm able to create the view , i'm able to populate it ( based on history ) , but as soon as i 'm pushing new data , the insertion fail with an unknown identifier ( which can be anything ).

To reproduce with a simplified version : 

Create the input table with source and destination 
```
CREATE TABLE  test.test_input (
  date_time DateTime('UTC'),
  site_src String,
  site_dst String,
  criticality UInt32,
  in_bytes Nullable(UInt64),
  out_bytes Nullable(UInt64)
) ENGINE = MergeTree PARTITION BY toStartOfDay(date_time)
ORDER BY
  (
    site_src,
    site_dst
  ) SETTINGS index_granularity = 8192
  
```
  
Then insert a few lines

 ```
insert into  test.test_input(date_time,site_src,site_dst,criticality,in_bytes,out_bytes) values 
     (toDateTime('2018-08-15 12:00:01'),'a','b',2,123,345),
	 (toDateTime('2018-08-15 12:00:01'),'c','b',2,123,345),
	 (toDateTime('2018-08-15 12:00:01'),'c','a',2,123,345),
	 (toDateTime('2018-08-15 12:00:01'),'b','c',2,123,345)

```
create a MV ( with populate ) converting src/dst into site. As you can see , i'm building an array based on source/destination to do a join and get the full list of site. I'm using arrayEnumerate to know in which direction i am.

```
create MATERIALIZED VIEW test.mv (
  date_time DateTime('UTC'),
  site String,
  lan_byte AggregateFunction(sum, UInt64),
  cnt AggregateFunction(count, UInt8)
) ENGINE = AggregatingMergeTree PARTITION BY toStartOfDay(date_time)
ORDER BY
  (
    site,
    date_time
  ) SETTINGS index_granularity = 8192 populate  AS
select
  date_time,
  site,
  sumState(if(idx = 1, in_bytes, out_bytes)) as lan_byte,
  countState(1) as cnt
from
  (
    select
      array(
        site_src,
        site_dst
      ) as sa,
      date_time ,
	  in_bytes,
	  out_bytes,
      if(idx==1,site_src,site_dst)  as site,
      idx
    from
      test.test_input ARRAY
      JOIN sa as keyS,
      arrayEnumerate(sa) as idx
    where
      (criticality >1
      )
  )
group by
  site,
  date_time
 
```
Here , everything is fine. The view contains accurate result and no error at all.
But when i'm trying to insert new data 
```
insert into  test.test_input(date_time,site_src,site_dst,criticality,in_bytes,out_bytes) values 
     (toDateTime('2018-08-15 12:00:01'),'a','b',2,123,345),
	 (toDateTime('2018-08-15 12:00:01'),'c','b',2,123,345),
	 (toDateTime('2018-08-15 12:00:01'),'c','a',2,123,345),
	 (toDateTime('2018-08-15 12:00:01'),'b','c',2,123,345)

```
I get the following error message : Code: 47, e.displayText() = DB::Exception: Unknown identifier: idx: while pushing to view test.mv, e.what() = DB::Exception

Did i do something wrong or unsupported ?
Is there a workaround ?

Thanks in advance 

The error in the server log is the following : 
```
2018.10.23 15:30:24.277967 [ 56 ] {e51d40a9-5a80-4f1a-8cc2-bb50e69788f5} <Error> executeQuery: Code: 47, e.displayText() = DB::Exception: Unknown identifier: idx: while pushing to view test.mv, e.what() = DB::Exception (from [::ffff:10.1.30.136]:64766) (in query:  /*TABIX_QUERY_ID_mkdqE0hF*/ ^M  ^M   insert into  test.test_input(date_time,site_src,site_dst,criticality,in_bytes,out_bytes) values), Stack trace:

0. /usr/bin/clickhouse-server(StackTrace::StackTrace()+0x16) [0x99e4866]
1. /usr/bin/clickhouse-server(DB::Exception::Exception(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, int)+0x22) [0x3084732]
2. /usr/bin/clickhouse-server(DB::ExpressionAnalyzer::collectUsedColumns()+0xb59) [0x7b1c259]
3. /usr/bin/clickhouse-server(DB::ExpressionAnalyzer::ExpressionAnalyzer(std::shared_ptr<DB::IAST> const&, DB::Context const&, std::shared_ptr<DB::IStorage> const&, DB::NamesAndTypesList const&, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&, unsigned long, bool, std::unordered_map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, DB::SubqueryForSet, std::hash<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::equal_to<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, DB::SubqueryForSet> > > const&)+0x888) [0x7b1ab98]
4. /usr/bin/clickhouse-server(DB::InterpreterSelectQuery::InterpreterSelectQuery(std::shared_ptr<DB::IAST> const&, DB::Context const&, std::shared_ptr<DB::IBlockInputStream> const&, std::shared_ptr<DB::IStorage> const&, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&, DB::QueryProcessingStage::Enum, unsigned long, bool)+0x443) [0x75f0693]
5. /usr/bin/clickhouse-server(DB::InterpreterSelectQuery::InterpreterSelectQuery(std::shared_ptr<DB::IAST> const&, DB::Context const&, std::shared_ptr<DB::IBlockInputStream> const&, DB::QueryProcessingStage::Enum, bool)+0x54) [0x75f1654]
6. /usr/bin/clickhouse-server(DB::PushingToViewsBlockOutputStream::process(DB::Block const&, unsigned long)+0xd3) [0x79edac3]
7. /usr/bin/clickhouse-server(DB::PushingToViewsBlockOutputStream::write(DB::Block const&)+0x8e) [0x79ee20e]
8. /usr/bin/clickhouse-server(DB::SquashingBlockOutputStream::writeSuffix()+0x13e) [0x79f635e]
9. /usr/bin/clickhouse-server(DB::copyData(DB::IBlockInputStream&, DB::IBlockOutputStream&, std::atomic<bool>*)+0x51a) [0x6af1faa]
10. /usr/bin/clickhouse-server(DB::executeQuery(DB::ReadBuffer&, DB::WriteBuffer&, bool, DB::Context&, std::function<void (std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)>)+0x1a3) [0x768f563]
11. /usr/bin/clickhouse-server(DB::HTTPHandler::processQuery(Poco::Net::HTTPServerRequest&, HTMLForm&, Poco::Net::HTTPServerResponse&, DB::HTTPHandler::Output&)+0x3169) [0x309fb39]
12. /usr/bin/clickhouse-server(DB::HTTPHandler::handleRequest(Poco::Net::HTTPServerRequest&, Poco::Net::HTTPServerResponse&)+0x3be) [0x30a215e]
13. /usr/bin/clickhouse-server(Poco::Net::HTTPServerConnection::run()+0x332) [0x9b6a712]
14. /usr/bin/clickhouse-server(Poco::Net::TCPServerConnection::start()+0xf) [0x9b62fbf]
15. /usr/bin/clickhouse-server(Poco::Net::TCPServerDispatcher::run()+0x16a) [0x9b6339a]
16. /usr/bin/clickhouse-server(Poco::PooledThread::run()+0x77) [0x9c80177]
17. /usr/bin/clickhouse-server(Poco::ThreadImpl::runnableEntry(void*)+0x38) [0x9c7bce8]
18. /usr/bin/clickhouse-server() [0xa3d5b4f]
19. /lib/x86_64-linux-gnu/libpthread.so.0(+0x76ba) [0x7f8e057ae6ba]
20. /lib/x86_64-linux-gnu/libc.so.6(clone+0x6d) [0x7f8e04dcf41d]

2018.10.23 15:30:24.278368 [ 56 ] {} <Error> HTTPHandler: Code: 47, e.displayText() = DB::Exception: Unknown identifier: idx: while pushing to view test.mv, e.what() = DB::Exception, Stack trace:

0. /usr/bin/clickhouse-server(StackTrace::StackTrace()+0x16) [0x99e4866]
1. /usr/bin/clickhouse-server(DB::Exception::Exception(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, int)+0x22) [0x3084732]
2. /usr/bin/clickhouse-server(DB::ExpressionAnalyzer::collectUsedColumns()+0xb59) [0x7b1c259]
3. /usr/bin/clickhouse-server(DB::ExpressionAnalyzer::ExpressionAnalyzer(std::shared_ptr<DB::IAST> const&, DB::Context const&, std::shared_ptr<DB::IStorage> const&, DB::NamesAndTypesList const&, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&, unsigned long, bool, std::unordered_map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, DB::SubqueryForSet, std::hash<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::equal_to<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, DB::SubqueryForSet> > > const&)+0x888) [0x7b1ab98]
4. /usr/bin/clickhouse-server(DB::InterpreterSelectQuery::InterpreterSelectQuery(std::shared_ptr<DB::IAST> const&, DB::Context const&, std::shared_ptr<DB::IBlockInputStream> const&, std::shared_ptr<DB::IStorage> const&, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&, DB::QueryProcessingStage::Enum, unsigned long, bool)+0x443) [0x75f0693]
5. /usr/bin/clickhouse-server(DB::InterpreterSelectQuery::InterpreterSelectQuery(std::shared_ptr<DB::IAST> const&, DB::Context const&, std::shared_ptr<DB::IBlockInputStream> const&, DB::QueryProcessingStage::Enum, bool)+0x54) [0x75f1654]
6. /usr/bin/clickhouse-server(DB::PushingToViewsBlockOutputStream::process(DB::Block const&, unsigned long)+0xd3) [0x79edac3]
7. /usr/bin/clickhouse-server(DB::PushingToViewsBlockOutputStream::write(DB::Block const&)+0x8e) [0x79ee20e]
8. /usr/bin/clickhouse-server(DB::SquashingBlockOutputStream::writeSuffix()+0x13e) [0x79f635e]
9. /usr/bin/clickhouse-server(DB::copyData(DB::IBlockInputStream&, DB::IBlockOutputStream&, std::atomic<bool>*)+0x51a) [0x6af1faa]
10. /usr/bin/clickhouse-server(DB::executeQuery(DB::ReadBuffer&, DB::WriteBuffer&, bool, DB::Context&, std::function<void (std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)>)+0x1a3) [0x768f563]
11. /usr/bin/clickhouse-server(DB::HTTPHandler::processQuery(Poco::Net::HTTPServerRequest&, HTMLForm&, Poco::Net::HTTPServerResponse&, DB::HTTPHandler::Output&)+0x3169) [0x309fb39]
12. /usr/bin/clickhouse-server(DB::HTTPHandler::handleRequest(Poco::Net::HTTPServerRequest&, Poco::Net::HTTPServerResponse&)+0x3be) [0x30a215e]
13. /usr/bin/clickhouse-server(Poco::Net::HTTPServerConnection::run()+0x332) [0x9b6a712]
14. /usr/bin/clickhouse-server(Poco::Net::TCPServerConnection::start()+0xf) [0x9b62fbf]
15. /usr/bin/clickhouse-server(Poco::Net::TCPServerDispatcher::run()+0x16a) [0x9b6339a]
16. /usr/bin/clickhouse-server(Poco::PooledThread::run()+0x77) [0x9c80177]
17. /usr/bin/clickhouse-server(Poco::ThreadImpl::runnableEntry(void*)+0x38) [0x9c7bce8]
18. /usr/bin/clickhouse-server() [0xa3d5b4f]
19. /lib/x86_64-linux-gnu/libpthread.so.0(+0x76ba) [0x7f8e057ae6ba]
20. /lib/x86_64-linux-gnu/libc.so.6(clone+0x6d) [0x7f8e04dcf41d]


```

**Comment by TH-HA** [comment]
I ve just realized that a 18.14.9 is available. I'll try it to see if some fixes are relevant for me.

**Comment by TH-HA** [comment]
switching to 18.14.9 don't change the behavior.

**Comment by TH-HA** [comment]
Fyi , i ve tried another view definition with the same result 

```
CREATE MATERIALIZED VIEW test.mv (
  date_time DateTime('UTC'),
  site String,
  lan_byte AggregateFunction(sum, UInt64),
  cnt AggregateFunction(count, UInt8)
) ENGINE = AggregatingMergeTree PARTITION BY toStartOfDay(date_time)
ORDER BY
  (site, date_time) SETTINGS index_granularity = 8192 AS
SELECT
  date_time,
  site,
  sumState(bytes) AS lan_byte,
  countState(1) AS cnt
FROM
  (
    SELECT
      [site_src, site_dst] AS sa,
      date_time,
      arrayJoin(sa) AS s,
      if(s = site_src, in_bytes, out_bytes) AS bytes,
      if(s = site_src, site_src, site_dst) AS site
    FROM
      test.test_input
    WHERE
      criticality > 1
  )
GROUP BY
  site,
  date_time
```

**Comment by TH-HA** [comment]
Seems to be related to #127 and #448 except it's on one single input table.  
Can we help to make progress on this issue ?  Materialized View combine with state are a killer feature , by solving this (especially since you can process MV concurrently ) the possibility are just enormous in term of automated and efficient processing.

**Comment by alex-zaitsev** [comment]
Could be the same as #2878

**Comment by knoguchi** [comment]
I tested this in ClickHouse version 20.3.4.10, official debian package.
Create the table, mv, and insert new data, then run the following query.


```
localhost :) select date_time, site, sumMerge(lan_byte) lan_byte, countMerge(cnt) as cnt from test.mv group by date_time, site;

SELECT
    date_time,
    site,
    sumMerge(lan_byte) AS lan_byte,
    countMerge(cnt) AS cnt
FROM test.mv
GROUP BY
    date_time,
    site

┌───────────date_time─┬─site─┬─lan_byte─┬─cnt─┐
│ 2018-08-15 12:00:01 │ b    │      813 │   3 │
│ 2018-08-15 12:00:01 │ c    │      591 │   3 │
│ 2018-08-15 12:00:01 │ a    │      468 │   2 │
└─────────────────────┴──────┴──────────┴─────┘
```

**Comment by den-crane** [comment]
this issue is about sub-queries in MV, fixed in 19.14.3.3
